File: src/main/java/algorithm/lc/InsertionSortList.java
package algorithm.lc;

public class InsertionSortList {

	public class ListNode {
		int val;
		ListNode next;

		ListNode(int x) {
			val = x;
			next = null;
		}
	}

	// O(1) space, O(n^2) time
	public class Solution {
		public ListNode insertionSortList(ListNode head) {
			if (head == null || head.next == null) {
				return head;
			}
			ListNode newHead = new ListNode(0);
			newHead.next = head;
			ListNode sortedHead = head;
			ListNode remainingHead = head.next; 
			head.next = null;

			while (remainingHead != null) { // while there is node not sorted
				ListNode sortedPrev = newHead;
				while (sortedPrev.next != null && sortedPrev.next.val < remainingHead.val) {
					sortedPrev = sortedPrev.next;
				}
				ListNode newRemainingHead = remainingHead.next;
				if (sortedPrev.next == null) { // append to the tail of sorted list
					remainingHead.next = null;
					sortedPrev.next = remainingHead;
				} else { // insert into sorted list
					remainingHead.next = sortedPrev.next;
					sortedPrev.next = remainingHead;
				}
				remainingHead = newRemainingHead;
			}
			return newHead.next;
		}
	}	
}

File: src/main/python/lc/insertion_sort_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
  # @param head, a ListNode
  # @return a ListNode
  def insertionSortList(self, head):
    if head is None or head.next is None:
      return head
    new_head = ListNode(0)
    new_head.next = head
    remaining_head = head.next
    head.next = None
    
    while remaining_head is not None:
      sorted_prev = new_head  # find the insertion position
      while sorted_prev.next is not None and sorted_prev.next.val < remaining_head.val:
        sorted_prev = sorted_prev.next
        
      new_remaining_head = remaining_head.next
      if sorted_prev.next is None:  # append to tail of sorted list
        sorted_prev.next = remaining_head
        remaining_head.next = None
      else:  # insert into sorted list
        remaining_head.next = sorted_prev.next
        sorted_prev.next = remaining_head
        
      remaining_head = new_remaining_head
      
    return new_head.next
